Bug 1408100 - Upgrade to Bootstrap 4

camd opened the pull request:
Transition from Bootstrap 3 to 4
[Bug 1408100](https://bugzilla.mozilla.org/show_bug.cgi?id=1408100)

Note: This is not pixel-perfect with Bootstrap 3.  To do that would have meant overriding a lot of Bootstrap 4.  I hope this is agree-able with everyone.  As it is, I still had to do quite a bit of pixel-tweaking to get it to look nice again

I tried to split this into pieces that would take some of the noise out of it.  The final 3 commits don't really stand on their own, so we could arguably squash those together.  But this split should help with the review a bit.

I must apologize, this has been kind of messy to get us back close to where we were.  We already had lots of pixel tweaks in our old CSS, and this is just some different pixel tweaks.  :)

imbstack commented:
Looks great! Notes from testing on stage:

* Action menu items like "Add new jobs" are only clickable on the words themselves, not across the entire `<li>` element. The dropdown does close however, so it ends up being a bit confusing.
* The top navbar seems to have some extra space between the bottom of the buttons and the bottom of the bar. Their appears to be extra padding or something when hovering for the highlight and clicking on them brings up a weird blue shadow that doesn't quite fit either.
* The 'Other job actions' menu seems to be inside an extra ul/div or something? Highlights don't reach the whole way across when you hover and it kinda looks strange in general?

Other than those nits, everything seems to work how I would expect it to w.r.t. actions!

wlach commented:
Some comments re perfherder:

* "Add more test data" button in graphs view doesn't have a hand icon. Maybe: https://stackoverflow.com/questions/41913646/bootstrap-button-no-hand-cursor
* New test chooser has some issues compared to old one. Notably:
 * labels (e.g. "Framework") look wrong (not bold anymore)
 * "Add test data" and close button positioning swapped in a non-standard way (close button should be on right, not left -- vice versa for title)
 * Form elements also seem larger than I would expect/need (compare old vs. new)
* Tooltips in comparison view (e.g. https://treeherder.mozilla.org/perf.html#/comparesubtest?originalProject=mozilla-inbound&originalRevision=1f1b4cb86b5892bccae047959b480ae6b863a7fe&newProject=mozilla-inbound&newRevision=1c4b4486947b3af2e53ac489eb8a2caadbe66729&originalSignature=054905c426e16aca2e90a056152d32e9f9616511&newSignature=054905c426e16aca2e90a056152d32e9f9616511&framework=1) lost their pointers and got overly large font sizes. Before/after:
<img width="233" alt="screen shot 2017-10-27 at 3 11 26 pm" src="https://user-images.githubusercontent.com/20569/32121113-31611256-bb29-11e7-90da-fcd9bfd5f432.png">
<img width="243" alt="screen shot 2017-10-27 at 3 11 04 pm" src="https://user-images.githubusercontent.com/20569/32121123-3bcbdd52-bb29-11e7-9b90-72b2643a6511.png">
* Spacing between tooltips in graph details looks off (it's overly compact).  Before/after:
<img width="312" alt="screen shot 2017-10-27 at 3 18 17 pm" src="https://user-images.githubusercontent.com/20569/32121386-6e4112ce-bb2a-11e7-9dec-1ac340354b1e.png">
<img width="318" alt="screen shot 2017-10-27 at 3 18 11 pm" src="https://user-images.githubusercontent.com/20569/32121321-2448d7ce-bb2a-11e7-9279-ea2e7506ec8a.png">
* (most critical) Can't scroll down at all in alerts view: https://treeherder-prototype.herokuapp.com/perf.html#/alerts?hideDwnToInv=1 (I can't really give feedback on alerts until this is fixed)

I know some of this might sound picky but I spent quite a bit of time both writing the original CSS and reviewing other people's PRs to make sure information was formatted in the clearest possible way and it would kind of suck to lose that. I'm not saying that things have to look exactly the same as before, but it would be good to at least preserve the spirit of the original.

edmorley commented on package.json:
```diff
@@ -20,9 +20,9 @@
     "angular-ui-router": "0.4.3",
-    "bootstrap": "3.3.7",
+    "angular1-ui-bootstrap4": "^2.4.22",
+    "bootstrap": "4.0.0-beta",
```
Beta 2 is out, let's use that :-)

```
$ npm view bootstrap dist-tags
{ latest: '3.3.7', next: '4.0.0-beta.2' }
```

https://blog.getbootstrap.com/2017/10/19/bootstrap-4-beta-2/

edmorley commented on package.json:
```diff
@@ -20,9 +20,9 @@
-    "angular-ui-bootstrap": "1.3.3",
     "angular-ui-router": "0.4.3",
-    "bootstrap": "3.3.7",
+    "angular1-ui-bootstrap4": "^2.4.22",
```
For consistency with the other entries (and to make it easier to see what needs updating, given lack of tools compatibility with reading yarn.lock for outdatedness), I think we should remove the `^`.

edmorley commented on package.json:
```diff
@@ -39,6 +39,7 @@
     "neutrino-preset-react": "4.2.3",
     "ngreact": "0.5.0",
     "numeral": "2.0.6",
+    "popper.js": "^1.12.5",
```
The caret can be removed here too (and adjust to 1.12.6)

edmorley commented on ui/entry-failureviewer.js:
```diff
@@ -19,8 +19,9 @@ require('angular-resource');
 require('angular-sanitize');
 require('angular-local-storage');
+require('popper.js/dist/umd/popper');
 require('bootstrap/dist/js/bootstrap');
```
Let's switch the bootstrap entry to use `import` and the format recommended here:
https://getbootstrap.com/docs/4.0/getting-started/webpack/#importing-javascript

Also, that page doesn't say we have to also `require('popper.js/dist/umd/popper')`, were things not working without that?

camd replied:
Yeah, you're right.  I took it out and it still worked.  I was hitting so very many problems earlier on that I think I'd tried that and hadn't taken it out yet...  :)  Thanks for catching that.

wlach commented:
@camd the pagination now looks ok in the alerts view and seems to work ok, though we lost the padding around it. I'm imagining you just hadn't gotten around to fixing that yet... (perhaps it's just a matter of making sure an existing css class is applied?)

wlach commented:
Actually, now that I look at it, there's many more issues with padding and spacing on the perfherder alerts view. A visual comparison between old/new should make them obvious.

camd commented:
I can see that a Selenium test failed.  I'll address that Monday.  :/

camd commented on ui/css/treeherder-resultsets.css:
```diff
@@ -187,3 +205,19 @@ fieldset[disabled] .btn-resultset:hover {
   background: #f8f8f8;
 }
+
+.get-next {
```
Bootstrap 4 does away with the concept of a ``well`` so this adds the styling to get it to look like that again.

wlach commented:
"runs" tooltip looks strange in new version (has a down arrow pointing at wrong spot). I'd say either center the arrow if possible, or just fall back to the tooltips without arrows. Example URL: https://treeherder-prototype.herokuapp.com/perf.html#/comparesubtest?originalProject=mozilla-inbound&originalRevision=209df98be4672182897eb92c2b7b807b2f7901b7&newProject=mozilla-inbound&newRevision=e9349ad2f1f8fec862b1d2271d0d8f25ad0814d4&originalSignature=f7a7eb10a31d949a8a02f1cee45a5ef452889bf7&newSignature=f7a7eb10a31d949a8a02f1cee45a5ef452889bf7&framework=1

<img width="129" alt="screen shot 2017-11-06 at 12 36 54 pm" src="https://user-images.githubusercontent.com/20569/32455115-50c2b906-c2ef-11e7-8d67-1e1dbf6bbd61.png">

Revisions text misaligned in alerts view:

<img width="491" alt="screen shot 2017-11-06 at 12 39 15 pm" src="https://user-images.githubusercontent.com/20569/32455208-8e1118a2-c2ef-11e7-850a-a04e31bb8233.png">

Link: https://treeherder-prototype.herokuapp.com/perf.html#/alerts?hideDwnToInv=1

wlach commented:
Not seeing hand cursor on close button on modals. E.g.:

<img width="153" alt="screen shot 2017-11-06 at 12 53 02 pm" src="https://user-images.githubusercontent.com/20569/32455895-ba873bb2-c2f1-11e7-91ee-e1dd3d2598aa.png">

wlach commented:
Tooltips also look misaligned/weird in alerts view. E.g.:

<img width="337" alt="screen shot 2017-11-06 at 12 58 15 pm" src="https://user-images.githubusercontent.com/20569/32456053-3911e5a4-c2f2-11e7-9e30-eae308f05281.png">

wlach commented on ui/css/perf.css:
```diff
@@ -576,7 +577,7 @@ li.pagination-next .material-icons::before {
   height: 0;
   position: absolute;
-  margin-left: 80px;
+  margin-left: 47%;
```
Fixed pixel values would be better than percentage ones here, it's easier to reason about what they're doing.

camd replied:
I needed to opt for a percentage because the popup could be a different height/width, depending on content.  Sometimes it has text of "Confidence value as calculated by Perfherder alerts.  Note that...."  and sometimes it's just "4 base / 4 new".  With the pixel value, the shorter popup has the arrow way off to the side.  This at least keeps it very close to the center.

wlach replied:
Oh, the arrow positioning is based on this value? That's frustrating. :(

camd replied:
Yeah, the ``angular1-ui-bootstrap4`` package removed the HTML element that was used for the arrow.  Maybe they found it hard and just removed it to make things easier on themselves.  Not sure.  So I had to get all fancy-pants with a CSS ``after`` element to put it back.  

What's the downside of doing it this way?

wlach replied:
How about we just kill the arrow for now, since a bunch of other elements don't have them anyway. We can always reintroduce them when/at the point that we port the compare view widget to react.

camd commented:
@wlach I addressed all the items you mentioned above.  Though not sure it'll still be deployed to prototype by the time you look.  I'll push it out there first thing tomorrow.

edmorley commented:
I notice both before and after this PR, Treeherder is using:
https://github.com/mozilla/treeherder/blob/master/ui/vendor/css/bootstrap-non-responsive.css

Support for non-responsive has been removed from Bootstrap 4:
> Non-responsive usage of Bootstrap is no longer supported.
(From https://getbootstrap.com/docs/4.0/migration/#global-changes)

It seems like we shouldn't be using that file any more?

Note also that the "taken from" comment URL in that file now 404s, the new link appears to be:
https://getbootstrap.com/docs/3.3/examples/non-responsive/non-responsive.css

...but the CSS seems to have been updated since our copy. See also these pages from the Bootstrap 3 docs for reference:
https://getbootstrap.com/docs/3.3/getting-started/#disable-responsive
https://getbootstrap.com/docs/3.3/examples/non-responsive/

The only mention Bootstrap 4 now has about responsiveness appears to be here:
https://getbootstrap.com/docs/4.0/getting-started/introduction/#responsive-meta-tag

...our `viewport` meta doesn't match what they suggest there, should it?

edmorley commented:
I ran bootlint using their bookmarklet against a loaded mozilla-inbound page via `yarn start` and it found a handful of errors. I'm not sure if any were pre-existing or not, but might be worth checking?
https://github.com/twbs/bootlint#in-the-browser

edmorley commented:
Hmm though looks like the current stable bootlint doesn't yet have official bootstrap 4 support, so perhaps not worth worrying about for now:
https://github.com/twbs/bootlint/pull/410

camd commented:
Thanks for catching the stuff about the non-responsive css.  I removed it.  I don't think it was doing much as things seemed the same with and without.   I *do* think we should go with the ``viewport`` setting they suggested.  I updated that as well.

I also removed the arrows from the popups as @wlach suggested.

edmorley commented:
> * the repos menu: some rows are no longer two columns wide, plus the links are a different colour and there is now a hover highlight per group (the latter two not a big deal, I'll leave that up to you)

Screenshot to explain this one a bit better :-)

![screenshot 3](https://user-images.githubusercontent.com/501702/32564167-43e3efb2-c4ab-11e7-876f-742f9a62bea5.png)

camd commented on ui/index.html:
```diff
@@ -3,7 +3,7 @@
         <meta charset="utf-8">
         <title ng-bind="getWindowTitle()">Treeherder</title>
-        <meta name="viewport" content="width=device-width, initial-scale=1.0">
+        <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
```
They weren't there before, so I hadn't added them.  But we might as well.

camd commented:
@edmorley : I pushed up the latest changes on prototype.  I tested this with the only windows VM I have: Windows 7...  :)  It looked good there and I didn't see the gaps you were describing.  I presume you're on windows 10, though.

edmorley commented:
It only occurs with narrow window sizes I think

camd commented on ui/entry-index.js:
```diff
@@ -7,7 +7,6 @@ require('./js/config');
 // Styles
 require('bootstrap/dist/css/bootstrap.css');
 require('font-awesome/css/font-awesome.css');
-require('./vendor/css/bootstrap-non-responsive.css');
```
ahh, yep.  removed

camd commented on neutrino-custom/base.js:
```diff
@@ -166,6 +166,7 @@ module.exports = neutrino => {
             jQuery: require.resolve('jquery'),
             'window.$': require.resolve('jquery'),
             'window.jQuery': require.resolve('jquery'),
+            Popper: ['popper.js', 'default'],
```
done

camd commented:
Yeah, I was trying with really narrow window sizes, kind of like the screenshots.  But didn't look wrong in the way your screenshot did.  :)  I ordered a copy of windows 10 so I'll have it for testing in a VM.  Not sure how long that will take...

edmorley commented:
I don't know if this is of any use?
https://developer.microsoft.com/en-us/windows/downloads/virtual-machines

Also, if it helps, I'm using Nightly.

edmorley commented:
Or this might be a smaller download?
https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/
